feature-libs/product/configurators/rulebased/root/interceptor/cpq-access-storage.service.ts

```diff
@@ -0,0 +1,114 @@
+import { Injectable } from '@angular/core';
+import { AuthService } from '@spartacus/core';
+import { BehaviorSubject, Observable, Subscription, timer } from 'rxjs';
+import { expand, filter, switchMap } from 'rxjs/operators';
+import { CpqAccessData } from './cpq-access-data.models';
+import { CpqAccessLoaderService } from './cpq-access-loader.service';
+
+export abstract class CpqConfiguratorTokenConfig {
+  cpqConfigurator: {
+    /* We should stop using/sending a token shortly before expiration,
+     * to avoid that it is actually expired when evaluated in the target system.
+     * Time given in ms. */
+    tokenExpirationBuffer: number;
+    /* max time in ms to pass until a token is considered expired and refetched,
+     * even if token expiration time is longer */
+    tokenMaxValidity: number;
+    /* min time to pass until a token is refecthed, even if token expiration time is shorter */
+    tokenMinValidity: number;
+  };
+}
+
+export const DefaultCpqConfiguratorTokenConfig: CpqConfiguratorTokenConfig = {
+  cpqConfigurator: {
+    tokenExpirationBuffer: 10000, // ten seconds
+    tokenMaxValidity: 24 * 60 * 60 * 1000, //one day
+    tokenMinValidity: 5000, // five seconds
+  },
+};
+
+const EXPIRED_TOKEN = {
+  accessToken: 'INVALID DUMMY',
+  accessTokenExpirationTime: 0,
+};
+
+@Injectable({ providedIn: 'root' })
+export class CpqAccessStorageService {
+  constructor(
+    protected cpqAccessLoaderService: CpqAccessLoaderService,
+    protected authService: AuthService,
+    protected config: CpqConfiguratorTokenConfig
+  ) {}
+
+  protected cpqAccessObservable: Observable<CpqAccessData>;
+  protected currentCpqAccessSubscription: Subscription;
+  protected cpqAccessDataCache: BehaviorSubject<CpqAccessData>;
+
+  getCachedCpqAccessData(): Observable<CpqAccessData> {
+    if (!this.cpqAccessObservable) {
+      this.initCpqAccessObservable();
+    }
+    return this.cpqAccessObservable;
+  }
+
+  protected initCpqAccessObservable() {
+    this.cpqAccessDataCache = new BehaviorSubject(EXPIRED_TOKEN);
+    this.cpqAccessObservable = this.cpqAccessDataCache.pipe(
+      // Never expose expired tokens - either cache was invalidated with expired token,
+      // or the cached one expired before a new one was fetched.
+      filter((data) => !this.isTokenExpired(data))
+    );
+    this.authService.isUserLoggedIn().subscribe((loggedIn) => {
```

Comment:
Even if technically not required: 
Let's add pipe(dinstinctUntilChanged), because authService.isUserLoggedIn() might change its behaviour, and in addition everyone, when viewing this code, needs to check that it does distinctUntilChanged internally..